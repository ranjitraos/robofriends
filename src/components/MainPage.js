import React, { useEffect } from 'react';
import CardList from '../components/CardList';
import SearchBox from '../components/SearchBox';
import Scroll from '../components/Scroll';
import ErrorBoundary from '../components/ErrorBoundary';
import Header from '../components/Header';
import './MainPage.css';

const MainPage = ({ searchField, onSearchChange, robots, isPending, onRequestRobots }) => {

    useEffect(() => {
        onRequestRobots()
    }, [onRequestRobots])

    const filterRobots = robots => {
        return robots.filter(robot => {
            return robot.name.toLowerCase().includes(searchField.toLowerCase());
        })
    }
    return isPending ?
        <h1>Loading</h1> :
        (
            <div className='tc'>
                <Header />
                <SearchBox searchChange={onSearchChange} />
                <Scroll>
                    <ErrorBoundary>
                        <CardList robots={filterRobots(robots)} />
                    </ErrorBoundary>
                </Scroll>
            </div>
        );
}


export default MainPage;