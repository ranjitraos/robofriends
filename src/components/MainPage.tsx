import React, { useEffect } from 'react';
import CardList from './CardList';
import SearchBox from './SearchBox';
import Scroll from './Scroll';
import ErrorBoundary from './ErrorBoundary';
import Header from './Header';
import './MainPage.css';

export interface IRobot {
    name: string;
    id: number;
    email: string;
}

interface IMainPageProps {
    searchField: string,
    onSearchChange(e: React.SyntheticEvent<HTMLInputElement>): void,
    robots: IRobot[],
    isPending: boolean,
    onRequestRobots: Function
}

const MainPage: React.FC<IMainPageProps> = ({ searchField, onSearchChange, robots, isPending, onRequestRobots }) => {

    useEffect(() => {
        onRequestRobots()
    }, [onRequestRobots])

    const filterRobots = (robots: IRobot[]) => {
        return robots.filter((robot: IRobot) => {
            return robot.name.toLowerCase().includes(searchField.toLowerCase());
        })
    }
    return isPending ?
        <h1>Loading</h1> :
        (
            <div className='tc'>
                <Header />
                <SearchBox searchChange={onSearchChange} />
                <Scroll>
                    <ErrorBoundary>
                        <CardList robots={filterRobots(robots)} />
                    </ErrorBoundary>
                </Scroll>
            </div>
        );
}


export default MainPage;