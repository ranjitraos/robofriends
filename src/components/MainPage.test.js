import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import MainPage from './MainPage';

describe('App component tests', () => {
    let wrapper;
    const mockProps = {
        onRequestRobots: jest.fn(),
        robots: [],
        searchField: "",
        isPending: false
    }
    beforeEach(() => {
        wrapper = renderer
            .create(<MainPage {...mockProps} />)
            .toJSON();
    })

    it('renders App component correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('filters robots correctly', async () => {
        const mockProps2 = {
            onRequestRobots: jest.fn(),
            robots: [{
                id: 3,
                name: 'jon',
                email: 'jon@gmail.com'
            }],
            searchField: "j",
            isPending: false
        }
        const { container } = render(<MainPage {...mockProps} />)
        const { container: container2 } = render(<MainPage {...mockProps2} />)
        expect(container.getElementsByClassName('card').length).toEqual(0)
        expect(container2.getElementsByClassName('card').length).toEqual(1)
    })

    it('filters robots correctly 2', async () => {
        const mockProps3 = {
            onRequestRobots: jest.fn(),
            robots: [{
                id: 3,
                name: 'jon',
                email: 'jon@gmail.com'
            }],
            searchField: "a",
            isPending: false
        }
        const { container } = render(<MainPage {...mockProps3} />)
        expect(container.getElementsByClassName('card').length).toEqual(0)
    })
})