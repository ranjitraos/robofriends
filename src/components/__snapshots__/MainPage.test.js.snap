// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component tests renders App component correctly 1`] = `
<div
  className="tc"
>
  <h1
    className="f1"
  >
    RoboFriends - 
    0.1.4
  </h1>
  <div>
    <button
      data-testid="counter"
      onClick={[Function]}
    >
      Count: 
      0
    </button>
  </div>
  <input
    aria-label="Search Robots"
    className="pa3 ba b--green bg-lightest-blue"
    placeholder="search robots"
    type="search"
  />
  <div
    style={
      Object {
        "border": "1px solid black",
        "height": "500px",
        "overflowY": "scroll",
      }
    }
  >
    <div />
  </div>
</div>
`;
